### PR TITLE
Fix: Importing a Badge with new BadgeType no longer fails; Closes #794

### LINKED ISSUE #794: Badge importing error
Trying to import 3 badges (talents) from Hackerspace to new Wolfmedia site. Error:
[badge_import.pdf](https://github.com/bytedeck/bytedeck/files/5861303/badge_import.pdf)

### PR DESCRIPTION
**Pull request by 9921622**
Please ensure you are familiar with our Pull Request Description expectations described here: https://www.pullrequest.com/blog/writing-a-great-pull-request-description/
### What?
Fixed an issue where if you import a badge from an exportee with a BadgeType the importer doesn't have. It will fail.

### Why?
See #794 

### How?
1. Added new importing field and dehydrate method for that field
image
image

2. added before import, which gets or creates the BadgeType referencing `badge_type_name` and using that to modify `badge_type`
image

Have to create a `BadgeType` before `save_instance` (and by extension `after_import`) because `BadgeType` is a required field. Therefore, im unable to copy the same method as how `campaigns` are imported with `quests`

### Testing?
Not sure about automated testing as our version of [`import_export`](https://django-import-export.readthedocs.io/en/2.0.2/search.html?q=test&check_keywords=yes&area=default) does not mention testing. And later versions use `tox.`

Only alternative was manual testing. What I did:
1. On one tenant, create 2 badges. `Badge (Success)` works on import in `develop` branch, while `Badge (Fail)` does not (as it references a new badge type)
image
2. export each Badge as a separate file
image
3. import each file on a new tenat.
image
image
4. result
image
image
image

### Review request
@tylerecouture

## Summary by CodeRabbit

- **New Features**
  - Improved badge handling during import operations with new fields: `badge_type_name`, `badge_type_sort`, and `badge_type_icon`.
  - Enhanced import functionality by generating badge types based on provided information.

### CODE REVIEW

**Comment by 9921622** [comment]
@tylerecouture review

**Review by tylerecouture — CHANGES_REQUESTED** [review]
We can't create the badge type if we don't have the rest of the badge type's model field info (icon, description, etc)

Can you look at how campaigns are handled on the quest side? It seems like a similar issue.  E.g. a quest might have a campaign on the export side, but not on the import side. Behaviour (and code as much as possible) should match.

**Comment by 9921622** [comment]
Only made `name`, 'fa_icon', and `sort_order` as the export variables for `BadgeType` as it was the only things you could add in the `BadgeType` view.

Also you can't create the `BadgeType` in `after_import` like in quest's import export, since `BadgeType` is a required field unlike `Category`

**Comment by 9921622** [comment]
@tylerecouture not sure why its failing flake8?

**Comment by tylerecouture** [comment]
If you click the failing action and then click the flake 8 step it will show you the log:

**Comment by 9921622** [comment]
that doesnt have anything to do with my code

**Comment by tylerecouture** [comment]
I don't think our GitHub action flake8 matches the local dev config for flake8, which is annoying, otherwise you could just rune "flake8 src" (which still picks up most issues)

**Comment by 9921622** [comment]
@tylerecouture flake8 is failing on code from 8-9 years ago

image
image

**Comment by tylerecouture** [comment]
The Flake8 may only check files that have changed, so maybe that file hasn't changed since we added the Flake8 Action

**Comment by 9921622** [comment]
https://github.com/bytedeck/bytedeck/issues/1622 fixed an issue with flake8 failing

**Review by tylerecouture — CHANGES_REQUESTED** [review]
Review comments left.  You have a lot of little methods without docstrings.  Please review the contributing guidelines, specifically: https://github.com/bytedeck/bytedeck/blob/develop/CONTRIBUTING.md#documentation

File: src/badges/admin.py [hunk 1]
```diff
@@ -45,6 +49,27 @@ def dehydrate_prereq_import_ids(self, quest):
                 prereq_import_ids += "&" + str(p.get_prereq().import_id)
         return prereq_import_ids
 
+    def dehydrate_badge_type_name(self, badge):
+        ''' Use the name instead of id.
+        We cannot guarantee exportee's ids are the same as import's ids
+        Use badge_type.name instead as it is unique.
+        '''
+        # without badge.id check will some times crash
```
[the remark this answers is not in the document]
Comment (9921622) [reply src/badges/admin.py:57]: using a print statement for `dehydrate_badge_type_name` and importing you get these logs
image
```
web-1  | dehydrate_badge_type_name
web-1  |         {'_state': <django.db.models.base.ModelState object at 0x7f2557795dc0>, 'id': None, 'name': '', 'xp': 0, 'datetime_created': None, 'datetime_last_edit': None, 'short_description': None, 'series_id': None, 'badge_type_id': None, 'icon': None, 'sort_order': None, 'active': True, 'import_id': UUID('a6d8a1a4-1133-4b31-9a3f-5d359c8ae529'), 'map_transition': False, '_django_version': '3.2.25'}
web-1  | dehydrate_badge_type_name
web-1  |         {'_state': <django.db.models.base.ModelState object at 0x7f2565030f70>, 'id': 8, 'name': 'New Badge (Fail)', 'xp': 0, 'datetime_created': datetime.datetime(2024, 6, 18, 22, 50, 21, 529084, tzinfo=<UTC>), 'datetime_last_edit': datetime.datetime(2024, 6, 18, 22, 50, 21, 529105, tzinfo=<UTC>), 'short_description': '', 'series_id': None, 'badge_type_id': 29, 'icon': <ImageFieldFile: None>, 'sort_order': None, 'active': True, 'import_id': '826574c7-9dfa-441d-a8fc-032bd0129a0e', 'map_transition': False}
```

It seems like dehydrate runs twice during import. But its unclear why as the docs dont seem to say anything about that.
In the first run, the badge object (using `badge.__dict__`) has `None` for its variables, but in the second run

Ill just explicitly check if `badge.id is None` and explain why

[the remark this answers is not in the document]
Comment (9921622) [reply src/badges/admin.py:57]: also adding to what i said. using only `if not badge.badge_type` will give me this error. Which is why im using `badge.id`
```
Badge matching query does not exist.
Traceback (most recent call last):
File "/usr/local/lib/python3.8/site-packages/import_export/resources.py", line 644, in import_data_inner
self.after_import(dataset, result, using_transactions, dry_run, **kwargs)
File "/app/src/badges/admin.py", line 128, in after_import
parent_badge = Badge.objects.get(import_id=import_id)
File "/usr/local/lib/python3.8/site-packages/django/db/models/manager.py", line 85, in manager_method
return getattr(self.get_queryset(), name)(*args, **kwargs)
File "/usr/local/lib/python3.8/site-packages/django/db/models/query.py", line 435, in get
raise self.model.DoesNotExist(
badges.models.Badge.DoesNotExist: Badge matching query does not exist.

Line number: 1 - Badge has no badge_type.
, New Type, , fa-handshake-o, 8, , New Badge (Fail), 0, 2024-06-12 13:38:34, 2024-06-12 15:16:37, , , 31, , , 1, 826574c7-9dfa-441d-a8fc-032bd0129a0e, 0
Traceback (most recent call last):
File "/usr/local/lib/python3.8/site-packages/import_export/resources.py", line 508, in import_row
diff = self.get_diff_class()(self, original, new)
File "/usr/local/lib/python3.8/site-packages/import_export/resources.py", line 167, in __init__
self.left = self._export_resource_fields(resource, instance)
File "/usr/local/lib/python3.8/site-packages/import_export/resources.py", line 188, in _export_resource_fields
return [resource.export_field(f, instance) if instance else "" for f in resource.get_user_visible_fields()]
File "/usr/local/lib/python3.8/site-packages/import_export/resources.py", line 188, in <listcomp>
return [resource.export_field(f, instance) if instance else "" for f in resource.get_user_visible_fields()]
File "/usr/local/lib/python3.8/site-packages/import_export/resources.py", line 680, in export_field
return method(obj)
File "/app/src/badges/admin.py", line 57, in dehydrate_badge_type_name
if not badge.badge_type:
File "/usr/local/lib/python3.8/site-packages/django/db/models/fields/related_descriptors.py", line 197, in __get__
raise self.RelatedObjectDoesNotExist(
badges.models.Badge.badge_type.RelatedObjectDoesNotExist: Badge has no badge_type.
```

File: src/badges/admin.py [hunk 2]
```diff
@@ -45,6 +49,27 @@ def dehydrate_prereq_import_ids(self, quest):
...
+        # since "badge.badge_type does not exist"
+        if badge.id is None or badge.badge_type is None:
+            return None
+        return str(badge.badge_type)
+
+    def dehydrate_badge_type_sort(self, badge):
+        if badge.id is None or badge.badge_type is None:
```
[the remark this answers is not in the document]
Comment (9921622) [reply src/badges/admin.py:64]: explained in the comment above https://github.com/bytedeck/bytedeck/pull/1619#discussion_r1645197216

File: src/badges/admin.py [hunk 3]
```diff
@@ -75,6 +100,30 @@ def generate_simple_prereqs(self, parent_object, data_dict):
...
+            )
+
+            row['badge_type'] = badge_type.id
+
+    def before_import_row(self, row, **kwargs):
+        # can create badge type here as it seems like the new badgetype will be saved only be saved
+        # after admin has accepted the import changes
+        self.generate_badge_type(row)
+
     def after_import(self, dataset, result, using_transactions, dry_run, **kwargs):
```
[the remark this answers is not in the document]
Comment (9921622) [reply src/badges/admin.py:127]: after_import is a `model.Resource` function and something i didnt add

File: src/badges/admin.py [hunk 4]
```diff
@@ -75,6 +100,30 @@ def generate_simple_prereqs(self, parent_object, data_dict):
...
+                defaults['sort_order'] = bt_sort
+
+            badge_type, _ = BadgeType.objects.get_or_create(
+                name=bt_name,
+                defaults=defaults,
+            )
+
+            row['badge_type'] = badge_type.id
+
+    def before_import_row(self, row, **kwargs):
```
[the remark this answers is not in the document]
Comment (9921622) [reply src/badges/admin.py:122]: before_import is a `model.Resource` function. Im not sure if its necessary to doc string, as none of the build in views are docstringed

File: src/badges/admin.py [hunk 5]
```diff
@@ -75,6 +100,30 @@ def generate_simple_prereqs(self, parent_object, data_dict):
                     # Create a new prereq to this quest
                     Prereq.add_simple_prereq(parent_object, prereq_object)
 
+    def generate_badge_type(self, row):
```
[the remark this answers is not in the document]
Comment (9921622) [reply src/badges/admin.py:103]:
```python
""" modifies the row "badge_type" to the correct BadgeType.id defined by
 + badge_type_name
 + badge_type_sort
 + badge_type_icon
If badge type doesn't exist, it creates it.
If staff cancels import anything created here will be rolled back automatically.
"""
```

File: src/badges/admin.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (9921622) [reply src/badges/admin.py:127]: ill just link it to the docs

File: src/badges/admin.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (9921622) [reply src/badges/admin.py:122]: ill just link it to the docs

**Review by tylerecouture — APPROVED** [review]

**Pull request merged**